From 0377bb3f43e851735b586252ea89d072cd1a9242 Mon Sep 17 00:00:00 2001
From: Roger Siddons <rsiddons@mythtv.org>
Date: Fri, 29 Dec 2017 13:50:53 -0500
Subject: [PATCH 04/22] Playback: Add Unfinished recording filter

Adds a new filter that shows recordings with a last played position
i.e. they've been partly watched

Ordered by last played position, with 'most progressed' at top.

Only implemented on MythCenter-Wide theme.

Refs #12809

Signed-off-by: Peter Bennett <pbennett@mythtv.org>
---
 mythtv/programs/mythfrontend/playbackbox.cpp    | 90 ++++++++++++++++---------
 mythtv/programs/mythfrontend/playbackbox.h      |  6 ++
 mythtv/themes/MythCenter-wide/recordings-ui.xml | 25 ++++---
 3 files changed, 80 insertions(+), 41 deletions(-)

diff --git a/mythtv/programs/mythfrontend/playbackbox.cpp b/mythtv/programs/mythfrontend/playbackbox.cpp
index 8150904..4bb6b95 100644
--- a/mythtv/programs/mythfrontend/playbackbox.cpp
+++ b/mythtv/programs/mythfrontend/playbackbox.cpp
@@ -396,8 +396,12 @@ PlaybackBox::PlaybackBox(MythScreenStack *parent, QString name,
       // Recording Group settings
       m_groupDisplayName(ProgramInfo::i18n("All Programs")),
       m_recGroup("All Programs"),
+      m_liveTvGroupName(tr("Live TV")),
+      m_liveTvGroupLabel(m_liveTvGroupName.toLower()),
       m_watchGroupName(tr("Watch List")),
       m_watchGroupLabel(m_watchGroupName.toLower()),
+      m_unfinishedGroupName(tr("Unfinished")),
+      m_unfinishedGroupLabel(m_unfinishedGroupName.toLower()),
       m_viewMask(VIEW_TITLES),
 
       // General m_popupMenu support
@@ -1632,6 +1636,7 @@ bool PlaybackBox::UpdateUILists(void)
 
     QMap<QString, QString> sortedList;
     QMap<int, QString> searchRule;
+    QMap<uint, ProgramInfo*> unfinished;
 
     m_programInfoCache.Refresh();
 
@@ -1732,10 +1737,8 @@ bool PlaybackBox::UpdateUILists(void)
 
             if (!isLiveTvGroup && isLiveTVProg && (m_viewMask & VIEW_LIVETVGRP))
             {
-                QString tmpTitle = tr("Live TV");
-                sortedList[tmpTitle.toLower()] = tmpTitle;
-                m_progLists[tmpTitle.toLower()].push_front(p);
-                m_progLists[tmpTitle.toLower()].setAutoDelete(false);
+                m_progLists[m_liveTvGroupLabel].push_front(p);
+                m_progLists[m_liveTvGroupLabel].setAutoDelete(false);
                 continue;
             }
 
@@ -1787,13 +1790,18 @@ bool PlaybackBox::UpdateUILists(void)
             {
                 m_watchlist.Add(p);
             }
+
+            if ((m_viewMask & VIEW_UNFINISHED) && p->GetProgressPercent() > 0)
+            {
+                unfinished.insert(p->GetProgressPercent(), p);
+            }
         }
     }
 
     if (sortedList.empty())
     {
-        LOG(VB_GENERAL, LOG_WARNING, LOC + "SortedList is Empty");
-        m_progLists[""];
+        LOG(VB_GENERAL, LOG_WARNING, LOC +
+            "No recordings selected for display (formerly SortedList is Empty)");
         m_titleList << "";
         m_playList.clear();
 
@@ -1869,12 +1877,25 @@ bool PlaybackBox::UpdateUILists(void)
         m_progLists[m_watchGroupLabel].setAutoDelete(false);
     }
 
-    m_titleList = QStringList("");
+    m_titleList = QStringList("");  // All Progs always at top
+
     if (m_progLists[m_watchGroupLabel].size() > 0)
         m_titleList << m_watchGroupName;
-    if ((m_progLists["livetv"].size() > 0) &&
-        (!sortedList.values().contains(tr("Live TV"))))
-        m_titleList << tr("Live TV");
+
+    if (!unfinished.isEmpty())
+    {
+        // Most progressed at top
+        foreach (ProgramInfo* p, unfinished)
+            m_progLists[m_unfinishedGroupLabel].push_front(p);
+
+        m_progLists[m_unfinishedGroupLabel].setAutoDelete(false);
+        m_titleList << m_unfinishedGroupName;
+    }
+
+    if (m_progLists[m_liveTvGroupLabel].size() > 0
+            && !sortedList.values().contains(m_liveTvGroupName))
+        m_titleList << m_liveTvGroupName;
+
     m_titleList << sortedList.values();
 
     // Populate list of recording groups
@@ -4882,18 +4903,6 @@ bool ChangeView::Create()
                 m_parentScreen, SLOT(toggleRecGroupView(bool)));
     }
 
-    // TODO Do we need two separate settings to determine whether the watchlist
-    //      is shown? The filter setting be enough?
-        checkBox = dynamic_cast<MythUICheckBox*>(GetChild("watchlist"));
-        if (checkBox)
-        {
-            if (m_viewMask & PlaybackBox::VIEW_WATCHLIST)
-                checkBox->SetCheckState(MythUIStateType::Full);
-            connect(checkBox, SIGNAL(toggled(bool)),
-                    m_parentScreen, SLOT(toggleWatchListView(bool)));
-        }
-    //
-
     checkBox = dynamic_cast<MythUICheckBox*>(GetChild("searches"));
     if (checkBox)
     {
@@ -4903,17 +4912,32 @@ bool ChangeView::Create()
                 m_parentScreen, SLOT(toggleSearchView(bool)));
     }
 
-    // TODO Do we need two separate settings to determine whether livetv
-    //      recordings are shown? Same issue as the watchlist above
-        checkBox = dynamic_cast<MythUICheckBox*>(GetChild("livetv"));
-        if (checkBox)
-        {
-            if (m_viewMask & PlaybackBox::VIEW_LIVETVGRP)
-                checkBox->SetCheckState(MythUIStateType::Full);
-            connect(checkBox, SIGNAL(toggled(bool)),
-                    m_parentScreen, SLOT(toggleLiveTVView(bool)));
-        }
-    //
+    checkBox = dynamic_cast<MythUICheckBox*>(GetChild("watchlist"));
+    if (checkBox)
+    {
+        if (m_viewMask & PlaybackBox::VIEW_WATCHLIST)
+            checkBox->SetCheckState(MythUIStateType::Full);
+        connect(checkBox, SIGNAL(toggled(bool)),
+                m_parentScreen, SLOT(toggleWatchListView(bool)));
+    }
+
+    checkBox = dynamic_cast<MythUICheckBox*>(GetChild("unfinished"));
+    if (checkBox)
+    {
+        if (m_viewMask & PlaybackBox::VIEW_UNFINISHED)
+            checkBox->SetCheckState(MythUIStateType::Full);
+        connect(checkBox, SIGNAL(toggled(bool)),
+                m_parentScreen, SLOT(toggleUnfinishedView(bool)));
+    }
+
+    checkBox = dynamic_cast<MythUICheckBox*>(GetChild("livetv"));
+    if (checkBox)
+    {
+        if (m_viewMask & PlaybackBox::VIEW_LIVETVGRP)
+            checkBox->SetCheckState(MythUIStateType::Full);
+        connect(checkBox, SIGNAL(toggled(bool)),
+                m_parentScreen, SLOT(toggleLiveTVView(bool)));
+    }
 
     checkBox = dynamic_cast<MythUICheckBox*>(GetChild("watched"));
     if (checkBox)
diff --git a/mythtv/programs/mythfrontend/playbackbox.h b/mythtv/programs/mythfrontend/playbackbox.h
index b01d13a..cda12ca 100644
--- a/mythtv/programs/mythfrontend/playbackbox.h
+++ b/mythtv/programs/mythfrontend/playbackbox.h
@@ -153,6 +153,7 @@ class PlaybackBox : public ScheduleCommon
         VIEW_WATCHLIST  =  0x0008,
         VIEW_SEARCHES   =  0x0010,
         VIEW_LIVETVGRP  =  0x0020,
+        VIEW_UNFINISHED =  0x0040,
         // insert new entries above here
         VIEW_WATCHED    =  0x8000
     } ViewMask;
@@ -264,6 +265,7 @@ class PlaybackBox : public ScheduleCommon
     void toggleCategoryView(bool setOn)  { toggleView(VIEW_CATEGORIES, setOn); }
     void toggleRecGroupView(bool setOn)  { toggleView(VIEW_RECGROUPS, setOn); }
     void toggleWatchListView(bool setOn) { toggleView(VIEW_WATCHLIST, setOn); }
+    void toggleUnfinishedView(bool setOn){ toggleView(VIEW_UNFINISHED, setOn); }
     void toggleSearchView(bool setOn)    { toggleView(VIEW_SEARCHES, setOn); }
     void toggleLiveTVView(bool setOn)    { toggleView(VIEW_LIVETVGRP, setOn); }
     void toggleWatchedView(bool setOn)   { toggleView(VIEW_WATCHED, setOn); }
@@ -437,8 +439,12 @@ class PlaybackBox : public ScheduleCommon
     QString             m_recGroup;
     QString             m_curGroupPassword;
     QString             m_newRecGroup;
+    QString             m_liveTvGroupName;
+    QString             m_liveTvGroupLabel;
     QString             m_watchGroupName;
     QString             m_watchGroupLabel;
+    QString             m_unfinishedGroupName;
+    QString             m_unfinishedGroupLabel;
     ViewMask            m_viewMask;
 
     // Watchlist support
diff --git a/mythtv/themes/MythCenter-wide/recordings-ui.xml b/mythtv/themes/MythCenter-wide/recordings-ui.xml
index 0abbb29..1857e74 100644
--- a/mythtv/themes/MythCenter-wide/recordings-ui.xml
+++ b/mythtv/themes/MythCenter-wide/recordings-ui.xml
@@ -513,29 +513,38 @@
         </textarea>
 
         <checkbox name="recgroups" from="basecheckbox">
-            <position>90,190</position>
+            <position>90,180</position>
         </checkbox>
 
         <textarea name="grouplabel" from="titlelabel">
-            <position>140,190</position>
+            <position>140,180</position>
             <value>Show Recording Groups</value>
         </textarea>
 
         <checkbox name="watchlist" from="basecheckbox">
-            <position>90,240</position>
+            <position>90,220</position>
         </checkbox>
 
         <textarea name="watchlabel" from="titlelabel">
-            <position>140,240</position>
+            <position>140,220</position>
             <value>Show Watch List</value>
         </textarea>
 
+        <checkbox name="unfinished" from="basecheckbox">
+            <position>90,260</position>
+        </checkbox>
+
+        <textarea name="unfinishedlabel" from="titlelabel">
+            <position>140,260</position>
+            <value>Show Unfinished</value>
+        </textarea>
+
         <checkbox name="searches" from="basecheckbox">
-            <position>90,290</position>
+            <position>90,300</position>
         </checkbox>
 
         <textarea name="searchlabel" from="titlelabel">
-            <position>140,290</position>
+            <position>140,300</position>
             <value>Show Searches</value>
         </textarea>
 
@@ -549,11 +558,11 @@
         </textarea>
 
         <checkbox name="watched" from="basecheckbox">
-            <position>90,390</position>
+            <position>90,380</position>
         </checkbox>
 
         <textarea name="watchedlabel" from="titlelabel">
-            <position>140,390</position>
+            <position>140,380</position>
             <value>Show Watched Programs</value>
         </textarea>
 
-- 
2.7.1

